store/store.go

```diff
@@ -537,14 +504,13 @@ func (s *Store[H]) flush(ctx context.Context, headers ...H) error {
 		}
 	}
 
-	// marshal and add to batch reference to the new head
+	// marshal and add to batch reference to the new head and tail
 	head := *s.contiguousHead.Load()
-	b, err := head.Hash().MarshalJSON()
-	if err != nil {
+	if err = writeHeaderHashTo(ctx, batch, head, headKey); err != nil {
 		return err
 	}
-
-	if err := batch.Put(ctx, headKey, b); err != nil {
+	tail := *s.tailHeader.Load()
+	if err = writeHeaderHashTo(ctx, batch, tail, tailKey); err != nil {
```

Comment:
I think I confused you and myself thinking that advanced header reference is saved outside of `flush`, while its not. So I reverted saving references here in `flush`